File: 0216-Combination Sum III/0216-Combination Sum III.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> result;
        if (n <= 0 || n > 45 || k <= 0)
        {
            return result;
        }
        
        vector<int> path;
        dfs(k, n, 1, 0, path, result);
        return result;
    }
    
private:
    void dfs(int k, int n, int start, int sum, vector<int>& path, vector<vector<int>>& result)
    {
        if (path.size() == k)
        {
            if (sum == n)
            {
                result.push_back(path);
            }
            
            return;
        }
        
        for (int i = start; i <= 9; ++i)
        {
            if (sum + i > n)
            {
                break;
            }
            
            path.push_back(i);
            dfs(k, n, i + 1, sum + i, path, result);
            path.pop_back();
        }
    }
};

File: 0201-0300/0216-Combination Sum III/0216-Combination Sum III.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        if n < 0 or n > 45 or k <= 0:
            return []
        result = []
        path = []
        def dfs(start, s):
            if len(path) == k:
                if s == n:
                    result.append(path[:])
                return
            for i in range(start, 10):
                if s + i > n:
                    break
                path.append(i)
                dfs(i + 1, s + i)
                path.pop()
        
        dfs(1, 0)
        return result
